### PR TITLE
Modernize NetWkstaUserEnum in smb scanner

### PR DESCRIPTION
**Pull request by NtAlexio2**
This PR modernize old style codes written for `NetWkstaUserEnum` request in `auxiliary/scanner/smb/smb_enumusers_domain` module. Also supports `RPORT` option.

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `use auxiliary/scanner/smb/smb_enumusers_domain`
- [x] Set options
- [x] `run`

Old style:
```
msf6 auxiliary(scanner/smb/smb_enumusers) > use auxiliary/scanner/smb/smb_enumusers_domain
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > set RHOST 192.168.1.78
RHOST => 192.168.1.78
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > set SMBUSER Administrator
SMBUSER => Administrator
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > set SMBPass Admin@123
SMBPass => Admin@123
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > run

Login Failed: Unable to negotiate SMB1 with the remote host: Not a valid SMB packet
[*] 192.168.1.78:445      - LAB\Administrator, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$, LAB\WIN-B3LRQS7Q431$
[+] 192.168.1.78:445      - Found user: LAB\Administrator
[*] 192.168.1.78:         - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/smb/smb_enumusers_domain) >
```

New style:
```
msf6 auxiliary(scanner/smb/smb_enumusers) > use auxiliary/scanner/smb/smb_enumusers_domain
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > set RHOST 192.168.1.78
RHOST => 192.168.1.78
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > set SMBUSER Administrator
SMBUSER => Administrator
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > set SMBPass Admin@123
SMBPass => Admin@123
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > run

Logged-on Users
===============

    Name              Domain  Other Domains  Logon Server
    ----              ------  -------------  ------------
    Administrator     LAB                    WIN-B3LRQS7Q431
    WIN-B3LRQS7Q431$  LAB

[*] 192.168.1.78: - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/smb/smb_enumusers_domain) >
```

**NOTE**: Tested on **Windows Server 2019 Build 1809** box

You can find MSDN documentation [here](https://learn.microsoft.com/en-us/windows/win32/api/lmwksta/nf-lmwksta-netwkstauserenum).

### CODE REVIEW

File: modules/auxiliary/scanner/smb/smb_enumusers_domain.rb [hunk 1]
```diff
@@ -24,191 +25,109 @@ def initialize
...
-        return
-      rescue ::Exception
-        print_line($!.to_s)
-      end
+    unless results.to_s.empty?
+      results_table = format_results(results)
+      results_table.rows = results_table.rows.uniq # Remove potentially duplicate entries from port 139 & 445
+      
+      print_line
+      print_line results_table.to_s
```
Comment (smcintyre-r7) [inline modules/auxiliary/scanner/smb/smb_enumusers_domain.rb:129]: Would you be opposed to keeping this all on one line like it was before?

## Old Output
```
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > run

[*] Using existing session 1
[*] MSFLAB\smcintyre, MSFLAB\smcintyre, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$, MSFLAB\DC$
[+] Found user: MSFLAB\smcintyre
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/smb/smb_enumusers_domain) >
```

## New Output
```
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > run


Logged-on Users
===============

    Name       Domain  Other Domains  Logon Server
    ----       ------  -------------  ------------
    DC$        MSFLAB
    smcintyre  MSFLAB                 DC

[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/smb/smb_enumusers_domain) >
```

## Proposed Output
I think ideally, if we kept the output on one line, uniqued the names, then comma joined them in a list enclosed in square brackets and prefixed with the domain, it'd be the same output as the `auxiliary/scanner/smb/smb_enumusers` module. It's worth us considering consolidating these two modules, but for now having output that looks the same would make it pretty easy to understand.

```
[+] MSFLAB [ smcintyre, DC$ ]
```

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
I left just a couple of comments. The only thing this is missing is that ruby_smb needs to be bumped to 3.3.10. You should be able to go into your `Gemfile.lock` and set the version from 3.3.9 to 3.3.10 to pull in your changes from ruby_smb.

File: modules/auxiliary/scanner/smb/smb_enumusers_domain.rb (hunk 1 above)
Comment (NtAlexio2) [reply modules/auxiliary/scanner/smb/smb_enumusers_domain.rb:129]: sure why not ^^

**Review by smcintyre-r7 — APPROVED** [review]
I made a small tweak to fix an exception when the connection failed. Everything else looks good, so I'll get this landed. Thanks for updating this, it's appreciated.

```
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > set VERBOSE true
VERBOSE => true
msf6 auxiliary(scanner/smb/smb_enumusers_domain) > run

[*] 192.168.159.10:445 - Connecting to Workstation Service Remote Protocol
[*] 192.168.159.10:445 - Binding to \wkssvc...
[+] 192.168.159.10:445 - Bound to \wkssvc
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10:445 - MSFLAB\DC$	(logon_server: , other_domains: )
[*] 192.168.159.10: - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/smb/smb_enumusers_domain) >
```

**Pull request merged**